### PR TITLE
 Technical design: scan install package of apps

### PR DESCRIPTION
**Pull request by zhou9584**
## Description
Design the architecture for APK Scanner. Considering the scalability, we would use some interference of task, result, scanner to implement multiple inheritance.

## Pull Request Checklist

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Code compiles correctly with all tests are passed.
- [x] I've read the [contributing guide](https://github.com/microsoft/HydraLab/blob/main/CONTRIBUTING.md#making-changes-to-the-code) and followed the recommended practices.
- [ ] [Wikis](https://github.com/microsoft/HydraLab/wiki) or [README](https://github.com/microsoft/HydraLab/blob/main/README.md) have been reviewed and added / updated if needed (for bug fixes / features)

### Does this introduce a breaking change?
*If this introduces a breaking change for Hydra Lab users, please describe the impact and migration path.*

- [ ] Yes
- [x] No

## How you tested it
*Please make sure the change is tested, you can test it by adding UTs, do local test and share the screenshots, etc.*

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [x] Technical design
- [ ] Build related changes
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Code style update (formatting, renaming) or Documentation content changes
- [ ] Other (please describe): 

### Feature UI screenshots or Technical design diagrams
*If this is a relatively large or complex change, kick it off by drawing the tech design with PlantUML and explaining why you chose the solution you did and what alternatives you considered, etc...*
image

### CODE REVIEW

File: agent/doc/Design/Installation Package Scanner Design.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+##### Data type
+
+![TaskDataType](../../../docs/images/UML/TaskDataType.png)
+
+##### Runtime sequence
+
+![package_scanner_design_sequence](../../../docs/images/UML/package_scanner_design_sequence.png)
+
+### Detailed Design
```
Comment (zhou9584) [inline agent/doc/Design/Installation Package Scanner Design.md:35]: disable agent type task

File: agent/doc/Design/Installation Package Scanner Design.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+1. Task registration
```
Comment (zhou9584) [inline agent/doc/Design/Installation Package Scanner Design.md:37]: excutor 兼容性检测 agent1: A  agent2:B  tas: AB

File: agent/doc/Design/Installation Package Scanner Design.md (hunk 1 above)
Comment (zhou9584) [inline agent/doc/Design/Installation Package Scanner Design.md:35]: 架构解决的问题列表，特性

**Comment by hydraxman** [comment]
Let's make sure empty desc do not appear going forward.

File: agent/doc/Design/Installation Package Scanner Design.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+3. Run task on Agent
+   1. When the Agent receives the AnalysisTask, it would choose a runner.
+   2. The runner would initialize a **AnalysisResult(APKReport)** object to save the scan result.
+   3. The runner would invoke multiple executors to prefer the APKReport object by the **AnalysisConfig** list of AnalysisTask.
+   4. The APKReport object would be transferred to the Center.
+4. Save to Database
+   1. There would be 3 tables to save task information: Task, TestTask, AnalysisTask with **InheritanceType.JOINED** strategy. The same property would be saved in the Task table. The unique info of TestTask and AnalysisTask would be saved in the other 2 tables.
+   2. For TaskResult, TestResult, AnalysisResult, is the same.
+5. Enable/disable AnalysisTask from Agent side
+   1. Add config in Agent side: **app.runner.analysis.enabled=true/false** to enable/disable the AnalysisTask.( default is false )
```
Comment (hydraxman) [inline agent/doc/Design/Installation Package Scanner Design.md:65]: This new line fixed the comment for "disable agent type task", @hydraxman

File: agent/doc/Design/Installation Package Scanner Design.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+The information includes the package name, version, and the list of files in the package. 
+BTW, the package scanner is also responsible for the package sensitivity verification.
+For example, whether any sensitive information such as secrets, tokens, sensitive words, etc. has been leaked.
+
+1. **Supported Scan Type**
+   1. Package information
+      1. [APKChecker](https://github.com/Tencent/matrix)
+   2. Package sensitivity
+      1. [APKLeaks](https://github.com/dwisiswant0/apkleaks)
+2. **Design Purpose**
```
Comment (zhou9584) [inline agent/doc/Design/Installation Package Scanner Design.md:17]: add this to fix '架构解决的问题列表，特性' comment

File: agent/doc/Design/Installation Package Scanner Design.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+### Detailed Design
+
+1. Task registration
+   1. The type of param that used to trigger task is still TestTaskSpec.
+   2. TestTaskSpec would be converted to AnalysisTask or TestTask by the property: **runningType**.
+2. Assign task to Agent
+   1. There would be a new property named as **availableAnalysisTaskCount** of "com.microsoft.hydralab.common.entity.center.AgentDeviceGroup" saved in Center memory.
+   2. When the Agent is online, the availableAnalysisTaskCount would be initialized by the **AgentFunctionAvailability**.
+   3. The Center would filter the Agent which has the **corresponding function** and **availableAnalysisTaskCount** >0 to assign the task.
```
Comment (zhou9584) [inline agent/doc/Design/Installation Package Scanner Design.md:55]: Add this to resolve 'excutor 兼容性检测'

**Review by hydraxman — APPROVED** [review]

**Pull request merged**